pkg/cli/cli.go

```diff
@@ -0,0 +1,498 @@
+			fmt.Printf(noticeColor, fmt.Sprintf("[Deprecation Notice] %s\n\n",
+				d.DeprecationWarning()))
+		}
+	}
+
+	return nil
+}
+
+// parseBaseFlags parses the command line arguments, looking for flags that
+// affect initialization of a cli. An error is returned only if an error
+// unrelated to flag parsing occurs.
+func (c *cli) parseBaseFlags() error {
+	// Create a dummy "base" flagset to populate from CLI args.
+	fs := pflag.NewFlagSet("base", pflag.ExitOnError)
+	fs.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
+
+	var help bool
+	// Set base flags that require pre-parsing to initialize c.
+	fs.BoolVarP(&help, helpFlag, "h", false, "print help")
+	fs.StringVar(&c.projectVersion, projectVersionFlag, c.defaultProjectVersion, "project version")
+	fs.StringVar(&c.cliPluginKey, pluginsFlag, "", "plugins to run")
+
+	// Parse current CLI args outside of cobra.
+	err := fs.Parse(os.Args[1:])
+	// User needs *generic* help if args are incorrect or --help is set and
+	// --project-version is not set. Plugin-specific help is given if a
+	// plugin.Context is updated, which does not require this field.
+	c.doGenericHelp = err != nil || help && !fs.Lookup(projectVersionFlag).Changed
+	c.cliPluginKey = strings.TrimSpace(c.cliPluginKey)
+
+	return nil
+}
+
+// validate validates fields in a cli.
+func (c cli) validate() error {
+	// Validate project versions.
+	if err := validation.ValidateProjectVersion(c.defaultProjectVersion); err != nil {
+		return fmt.Errorf("failed to validate default project version %q: %v", c.defaultProjectVersion, err)
+	}
+	if err := validation.ValidateProjectVersion(c.projectVersion); err != nil {
+		return fmt.Errorf("failed to validate project version %q: %v", c.projectVersion, err)
+	}
+
+	if _, versionFound := c.pluginsFromOptions[c.projectVersion]; !versionFound {
+		return fmt.Errorf("no plugins for project version %q", c.projectVersion)
+	}
+	// If --plugins is not set, no layout exists (no config or project is v1 or v2),
+	// and no defaults exist, we cannot know which plugins to use.
+	if (!c.configured || c.projectVersion != config.Version3) && c.cliPluginKey == "" {
+		_, versionExists := c.defaultPluginsFromOptions[c.projectVersion]
+		if !versionExists {
+			return fmt.Errorf("no default plugins for project version %s", c.projectVersion)
+		}
+	}
+
+	// Validate plugin versions and name.
+	for _, versionedPlugins := range c.pluginsFromOptions {
+		if err := validatePlugins(versionedPlugins...); err != nil {
+			return err
+		}
+	}
+	for _, defaultPlugins := range c.defaultPluginsFromOptions {
+		if err := validatePlugins(defaultPlugins...); err != nil {
+			return err
+		}
+	}
+
+	// Validate plugin keys set in CLI.
+	if c.cliPluginKey != "" {
+		pluginName, pluginVersion := plugin.SplitKey(c.cliPluginKey)
+		if err := plugin.ValidateName(pluginName); err != nil {
+			return fmt.Errorf("failed to validate plugin name %q: %v", pluginName, err)
```

Comment:
Minor nit for clarity: did we fail to validate or did the validation tell us the thing was invalid?
```suggestion
			return fmt.Errorf("invalid plugin name %q: %v", pluginName, err)
```